lib/filterx/expr-json.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2023 Balazs Scheidler <balazs.scheidler@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "expr-json.h"
+#include "object-json.h"
+#include "scratch-buffers.h"
+#include <json-c/json.h>
+
+struct _FilterXJSONKeyValue
+{
+  gchar *key;
+  FilterXExpr *value_expr;
+};
+
+FilterXJSONKeyValue *
+filterx_json_kv_new(const gchar *key, FilterXExpr *value_expr)
+{
+  FilterXJSONKeyValue *self = g_new0(FilterXJSONKeyValue, 1);
+  self->key = g_strdup(key);
+  self->value_expr = value_expr;
+  return self;
+}
+
+void
+filterx_json_kv_free(FilterXJSONKeyValue *self)
+{
+  g_free(self->key);
+  filterx_expr_unref(self->value_expr);
+  g_free(self);
+}
+
+typedef struct _FilterXJSONExpr
+{
+  FilterXExpr super;
+  GList *key_values;
+} FilterXJSONExpr;
+
+static gboolean
+_eval_key_value(FilterXJSONExpr *self, struct json_object *object, FilterXJSONKeyValue *kv)
+{
+  FilterXObject *value = filterx_expr_eval_typed(kv->value_expr);
+  gboolean success = FALSE;
+
+  if (!value)
+    goto fail;
+
+  struct json_object *value_json = NULL;
+  if (!filterx_object_map_to_json(value, &value_json))
+    goto fail;
+
+  /* don't store the object in the JSON cache in case it is JSON, as that
+   * would cause a circular reference */
+  if (!filterx_object_is_type(value, &FILTERX_TYPE_NAME(json)))
+    filterx_json_associate_cached_object(value_json, value);
+  if (json_object_object_add(object, kv->key, value_json) != 0)
+    {
+      json_object_put(value_json);
+      goto fail;
+    }
+  success = TRUE;
+fail:
+  filterx_object_unref(value);
+  return success;
+}
+
+static FilterXObject *
+_eval(FilterXExpr *s)
+{
+  FilterXJSONExpr *self = (FilterXJSONExpr *) s;
+  struct json_object *object;
+
+  object = json_object_new_object();
+  for (GList *l = self->key_values; l; l = l->next)
+    {
+      if (!_eval_key_value(self, object, l->data))
+        goto fail;
```

Comment:
The following evaluates to `false` (c is unset):

```
filterx{
  ${d} = {
    "c": $c,
    "d": 1,
  };
};
```

I'm not entirely sure this is an intuitive behavior, as a single missing value in a big key-value structure should not make syslog-ng drop the entire message.

Reply:
This is actually an "exception". If a variable is unset, I'd consider that an error which should not allow the filterx expression to continue.

The issue is that these kind of things are not reported properly (whereas they should).

So NULL values during evaluation is an "error" case, the problem is that the error condition itself is not recorded, thus not reported to the user either.

But this should become something like: "undefined variable $c at line:column" or something like that.

Reply:
We've discussed that the error reporting code is still missing and that to include a potentially unset value we have a couple of options:

1) use a template, e.g. instead of a straight $c, let's use "$c", the template would expand that to an empty string and the evaluation would continue

2) potentially add a tertiary expression, e.g. ?: and a conditional to check if a variable is set